<a href="https://colab.research.google.com/github/emanuel379/Analise-de-Sobrevivencia-e-Confiabilidade/blob/main/Mini_avalia%C3%A7%C3%A3o_5.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [ ]:
# Instalando as bibliotecas necessárias
install.packages(c("broom", "tidyverse", "pammtools", "numDeriv","gamlss", "gamlss.cens", "RelDists", "KMsurv", "survival", "survminer", "stats4"))

Installing packages into ‘/usr/local/lib/R/site-library’
(as ‘lib’ is unspecified)

Warning message:
“package ‘stats4’ is a base package, and should not be updated”
also installing the dependencies ‘listenv’, ‘parallelly’, ‘dotCall64’, ‘minqa’, ‘nloptr’, ‘shape’, ‘viridis’, ‘MatrixModels’, ‘TH.data’, ‘sandwich’, ‘htmlwidgets’, ‘RcppEigen’, ‘future’, ‘globals’, ‘spam’, ‘BH’, ‘carData’, ‘abind’, ‘pbkrtest’, ‘lme4’, ‘diagram’, ‘iterators’, ‘Hmisc’, ‘quantreg’, ‘SparseM’, ‘polspline’, ‘multcomp’, ‘htmlTable’, ‘cmprsk’, ‘doParallel’, ‘mets’, ‘plotrix’, ‘Publish’, ‘ranger’, ‘RcppArmadillo’, ‘future.apply’, ‘progressr’, ‘SQUAREM’, ‘orthopolynom’, ‘fields’, ‘mapproj’, ‘maps’, ‘MBA’, ‘corrplot’, ‘car’, ‘markdown’, ‘png’, ‘jpeg’, ‘prodlim’, ‘foreach’, ‘rms’, ‘riskRegression’, ‘lava’, ‘timereg’, ‘DEoptim’, ‘GA’, ‘gaussquad’, ‘autoimage’, ‘rbibutils’, ‘Rcpp’, ‘RcppParallel’, ‘ggrepel’, ‘ggsci’, ‘cowplot’, ‘ggsignif’, ‘polynom’, ‘rstatix’, ‘exactRankTests’, ‘zoo’, ‘km.ci’, ‘xtable’, ‘gridtext’, ‘ch

In [ ]:
# Carregando pacotes ------------------------------------------------------
# ---

library(gamlss)
library(gamlss.cens)
library(broom)
library(tidyverse)
library(pammtools)
library(numDeriv)
library(RelDists)
library(KMsurv)
library(survival)
library(survminer)
library(stats4)

Loading required package: splines

Loading required package: gamlss.data


Attaching package: ‘gamlss.data’


The following object is masked from ‘package:datasets’:

    sleep


Loading required package: gamlss.dist

Loading required package: nlme

Loading required package: parallel

 **********   GAMLSS Version 5.4-22  ********** 

For more on GAMLSS look at https://www.gamlss.com/

Type gamlssNews() to see new features/changes/bug fixes.


Loading required package: survival

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.1
✔ ggplot2   3.4.4     ✔ tibble    3.2.1
✔ lubridate 1.9.3     ✔ tidyr     1.3.1
✔ purrr     1.0.2     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ purrr::%||%()     masks base::%||%()
✖ dplyr::collapse() masks nlme::collapse()
✖ dplyr::filter()   masks stats::filter()
✖ dplyr::lag()      masks stats::lag()
ℹ Use the

In [ ]:
# Dados

data('std')

temp = std$time

cens = std$rinfct

yschool = std$yschool

npartner = std$npartner

abdpain = std$abdpain

dados = tibble(tempos=temp, cens=cens, yschool=yschool, npartner=npartner, abdpain=abdpain)

**1)** Ajuste um modelo de regressão de Weibull utilizando as seguintes covariáveis: **yschool** - Anos de estudo; **npartner** - Número de parceiros nos últimos 30 dias. Analise os resultados obtidos e classifique as alternativas. Obs: as covariáveis são incluídas no parâmetro **μ**
.

In [ ]:
# Estimacao de maxima verossimilhanca (com o pacote gamlss) ---------------
# ---
gen.cens(WEI)
mod1 = gamlss(formula=Surv(tempos,cens)~ yschool + npartner, family= cens(WEI), data=dados)
summary(mod1)

A censored family of distributions from WEI has been generated 
 and saved under the names:  
 dWEIrc pWEIrc qWEIrc WEIrc 
The type of censoring is right  
GAMLSS-RS iteration 1: Global Deviance = 5369.856 
GAMLSS-RS iteration 2: Global Deviance = 5369.772 
GAMLSS-RS iteration 3: Global Deviance = 5369.767 
GAMLSS-RS iteration 4: Global Deviance = 5369.766 
******************************************************************
Family:  c("WEIrc", "right censored Weibull") 

Call:  gamlss(formula = Surv(tempos, cens) ~ yschool + npartner,  
    family = cens(WEI), data = dados) 

Fitting method: RS() 

------------------------------------------------------------------
Mu link function:  log
Mu Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)  4.73256    0.49018   9.655  < 2e-16 ***
yschool      0.20881    0.04373   4.776  2.1e-06 ***
npartner    -0.07160    0.06548  -1.094    0.274    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

--------

I) Após o ajuste do modelo, identificou-se que as covariáveis **yschool** e **npartner** são significativas.

**Falsa**: uma vez que, considerando um nível de significância de 5%, temos que a variável **yschool** é, de fato, significativa para o modelo, mas **npartner** não.

II) A estimativa obtida para a covariável **yschool** foi de 0.20882 com erro padrão de 0.04373. E a estimativa obtida para a covariável **npartner** foi de -0.07160 com erro padrão de 0.06548.

**Verdadeira:** uma vez que os valores obtidos para as estimativas e o erro padrão para ambas as covariáveis com a tabela ANOVA são iguais aos apresentados na afirmativa.

III) Após o ajuste do modelo, identificou-se que apenas a covariável **yschool** é significativa.

**Verdadeira**: como já observado no item I.

IV) Todas as estimativas obtidas neste modelo ajustado podem ser interpretadas.

**Falsa**: uma vez que o intercepto é não interpretável.

**Resposta**:

c) Apenas II e III

**2)** Ajuste um modelo de regressão de Weibull utilizando a covariável **abdpain** - Presença de dor abdominal (1=sim e 0=não). Analise os resultados obtidos e selecione a alternativa correta. Obs: a covariável é incluida no parâmetro **μ**
.

In [ ]:
# Estimacao de maxima verossimilhanca (com o pacote gamlss) ---------------
# ---
gen.cens(WEI)
mod2 = gamlss(formula=Surv(tempos,cens)~ abdpain, family= cens(WEI), data=dados)
summary(mod2)

A censored family of distributions from WEI has been generated 
 and saved under the names:  
 dWEIrc pWEIrc qWEIrc WEIrc 
The type of censoring is right  
GAMLSS-RS iteration 1: Global Deviance = 5390.44 
GAMLSS-RS iteration 2: Global Deviance = 5390.309 
GAMLSS-RS iteration 3: Global Deviance = 5390.301 
GAMLSS-RS iteration 4: Global Deviance = 5390.3 
******************************************************************
Family:  c("WEIrc", "right censored Weibull") 

Call:  gamlss(formula = Surv(tempos, cens) ~ abdpain, family = cens(WEI),  
    data = dados) 

Fitting method: RS() 

------------------------------------------------------------------
Mu link function:  log
Mu Coefficients:
            Estimate Std. Error t value Pr(>|t|)    
(Intercept)   7.0634     0.0855  82.612   <2e-16 ***
abdpain      -0.3929     0.1933  -2.033   0.0424 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

------------------------------------------------------------------
Sigma li

a) As estimativas obtidas da modelagem são **μ1**=
 7.063, **μ2**=
-0.393, **σ**=
0.281

**Falsa**: uma vez que o valor de **σ** predito é -0.281.

b) A interpretação do parâmetro associado à variável **abdpain** é:  à medida que a dor abdominal aumenta em uma unidade, espera-se que o tempo de sobrevivência do paciente diminua em um fator de **e−0.393≈0.675**
.

**Verdadeira**: uma vez que se trata de uma transformação de variáveis via método Delta.

c) A matriz de variância-covariância é  

$$Cov=\begin{equation*}
\begin{bmatrix}
0.007310457 & 0.0063106395 & −0.0016065125 \\
0.006310640 & 0.0373636463 & −0.0003651316\\
−0.001606512 & −0.0003651316 & 0.0019946585
\end{bmatrix}
\end{equation*}$$

In [ ]:
vcov(mod2)

(Intercept)  abdpain       (Intercept)  
(Intercept)  0.007310445 -0.0063106166 -0.0016065086
abdpain     -0.006310617  0.0373637297  0.0003651106
(Intercept) -0.001606509  0.0003651106  0.0019946581

**Falsa**: uma vez que os valores a23 = a32 = 0.0003651316, enquanto na afirmativa esses aparecem como negativos.

d) O parâmetro **σ**
 representa o logaritmo do desvio padrão da distribuição Weibull.

**Falsa**: **σ** é um parâmetro estimado sem, necessáriamente, assumir qualquer relação com o desvio padrão.

e) A estimativa do **σ**
 indica que a forma da função de risco associada a esta distribuição está diminuindo ao longo do tempo.

 **Verdadeira:** Valores de **σ** menores do que 1 indicam que este impacta de forma a reduzir o risco.

**Resposta**:

a) FVFFV

**3)** Ajuste o modelo de Cox e considere as cováriaveis: yschool - Anos de estudo; npartner - Número de parceiros nos últimos 30 dias e abdpain - Presença de dor abdominal (1=sim e 0=não).Interprete seus parâmetros acerca da razão de riscos e julgue as sentenças abaixo.

In [ ]:
# Ajustando o modelo ------------------------------------------------------
# ---
fit3 = coxph(Surv(tempos,cens) ~ yschool + npartner + abdpain, data=dados, x = T)
summary(fit3)

Call:
coxph(formula = Surv(tempos, cens) ~ yschool + npartner + abdpain, 
    data = dados, x = T)

  n= 877, number of events= 347 

             coef exp(coef) se(coef)      z Pr(>|z|)    
yschool  -0.16328   0.84935  0.03324 -4.912 9.02e-07 ***
npartner  0.04937   1.05061  0.04990  0.989   0.3225    
abdpain   0.30038   1.35038  0.14675  2.047   0.0407 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

         exp(coef) exp(-coef) lower .95 upper .95
yschool     0.8494     1.1774    0.7958    0.9065
npartner    1.0506     0.9518    0.9527    1.1586
abdpain     1.3504     0.7405    1.0128    1.8004

Concordance= 0.575  (se = 0.017 )
Likelihood ratio test= 29.02  on 3 df,   p=2e-06
Wald test            = 28.73  on 3 df,   p=3e-06
Score (logrank) test = 28.58  on 3 df,   p=3e-06


I) O coeficiente negativo indica que uma diminuição no tempo de escolaridade está associado a uma redução no risco de reinfecção. O hazard ratio de 0.84945 significa que para cada unidade adicional no tempo de escolaridade, o risco de reinfecção diminui em aproximadamente 15%. Com um p-valor menor que 0.01, esta covariável é estatisticamente significativa, indicando que tempo de escolaridade é um fator relevante para o risco de reinfecção.

**Falsa:** a negatividade do coeficiente indica que um aumento no tempo de escolaridade reduz o risco de reinfecção, ou seja, se diminuirmos este tempo, acabamos por aumentar esse risco, corroborando a afirmação acerca do hazard ratio, que é verdadeira.

II) O coeficiente positivo indica que um aumento no número de parceiros está associado a um aumento no risco de reinfecção. O hazard ratio de 1.05059 significa que para cada parceiro adicional, o risco de reinfecção aumenta em aproximadamente 5%. No entanto, esta covariável não é estatisticamente significativa no modelo, sugerindo que o número de parceiros pode não ter um impacto substancial no risco de reinfecção.

**Verdadeira:** considerando que o valor-p é maior que o tradicional nível 0.05 de significância, temos que a covariável pode ser considerada não significativa, as interpretações acerca do coeficiente e do risco também estão corretas.

III) O coeficiente positivo indica que a presença de dor abdominal está associada a um aumento no risco de reinfecção. O hazard ratio de 1.34956 significa que a presença de dor abdominal aumenta o risco de reinfecção em aproximadamente 35%. Com um p-valor de 0.0411, esta covariável é estatisticamente significativa, indicando que a dor abdominal é um fator relevante para o risco de reinfecção.

**Verdadeira:** assim como no item precedente, temos que as interpretações para o coeficiente e o risco estão corretas, bem como o valor-p menor do que o nível de significância nos traz evidências acerca da significância estatística da covariável.

Resposta:

a) Apenas a afirmativa I é falsa.

**4)** Ajuste dois modelos de cox, primeiro considere as cováriaveis: npartner e abdpain. No segundo modelo adicione a interação entre as covariáveis “npartner” e “abdpain”. Acerca das afirmações abaixo classifique como verdadeira ou falsa.

In [ ]:
# Ajustando o modelo 1
fit41 = coxph(Surv(tempos,cens) ~ npartner + abdpain, data=dados, x = T)
summary(fit41)

Call:
coxph(formula = Surv(tempos, cens) ~ npartner + abdpain, data = dados, 
    x = T)

  n= 877, number of events= 347 

            coef exp(coef) se(coef)     z Pr(>|z|)  
npartner 0.04343   1.04438  0.04911 0.884   0.3766  
abdpain  0.29196   1.33905  0.14671 1.990   0.0466 *
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

         exp(coef) exp(-coef) lower .95 upper .95
npartner     1.044     0.9575    0.9485     1.150
abdpain      1.339     0.7468    1.0044     1.785

Concordance= 0.519  (se = 0.016 )
Likelihood ratio test= 4.66  on 2 df,   p=0.1
Wald test            = 5.04  on 2 df,   p=0.08
Score (logrank) test = 5.06  on 2 df,   p=0.08


In [ ]:
# Ajustando o modelo 2
fit42 = coxph(Surv(tempos,cens) ~ npartner + abdpain + npartner*abdpain, data=dados, x = T)
summary(fit42)

Call:
coxph(formula = Surv(tempos, cens) ~ npartner + abdpain + npartner * 
    abdpain, data = dados, x = T)

  n= 877, number of events= 347 

                      coef exp(coef)  se(coef)     z Pr(>|z|)
npartner         0.0002588 1.0002588 0.0660843 0.004    0.997
abdpain          0.1506696 1.1626124 0.1959681 0.769    0.442
npartner:abdpain 0.0990160 1.1040840 0.0899701 1.101    0.271

                 exp(coef) exp(-coef) lower .95 upper .95
npartner             1.000     0.9997    0.8787     1.139
abdpain              1.163     0.8601    0.7918     1.707
npartner:abdpain     1.104     0.9057    0.9256     1.317

Concordance= 0.519  (se = 0.016 )
Likelihood ratio test= 5.79  on 3 df,   p=0.1
Wald test            = 7.33  on 3 df,   p=0.06
Score (logrank) test = 7.76  on 3 df,   p=0.05


I) Todas as covariáveis no primeiro modelo foram significativas a um nível de 10% de significância, mas ao considerar um nível de 5% a covariável “abdpain” não é significativa.

**Falsa:** a covariável abdpain é significativa mesmo a um nível de significância de 5%, enquanto a covariável npartner pode ser considerada não significativa mesmo a um nível de significância de 10%.

II) O coeficiente estimado para a covariável “npartner” no primeiro modelo é positivo mas abaixo de 0.05, indicando que conforme o valor dela aumenta, o risco de falha diminui.

**Falsa**: uma vez que o hazard ratio de 1.04438 nos indica que para cada parceiro adicional, o risco de reinfecção aumenta em aproximadamente 4.4%.

III) A magnitude do coeficiente mostra a força da associação entre a covariável e o risco de falha. Coeficientes com maior valor absoluto têm um impacto mais significativo na função de risco. Com isso a covariável que indica dor abdominal é a que possui maior impacto no primeiro modelo de cox ajustado.

**Verdadeira:** uma vez que valores absolutos maiores dos coeficientes indicam maior impacto de sua respectiva covariável sobre a função de risco, além disso, com um hazard ratio de 1.33905, temos que a presença de dor abdominal aumenta o risco de reinfecção em cerca de 33.9%, sendo abdpain, por tanto, a covariável de maior impacto para o modelo 1.

IV) Ao ajustar o modelo que considera a interação entre as covariáveis “npartner” e “abdpain” verificamos que esta é a que possui menor p-valor, com isso é a mais significativa, portanto, sendo necessária considerá-la na modelagem final.

**Falsa:** apesar de possuir o menor valor-p, em 0.271 este ainda é muito superior às significâncias usuais, sendo assim, tal covariável, assim como as demais, também pode ser considerada não significativa estatisticamente.

Resposta:

a) FFVF

**5)** Ajuste dois modelos de cox, sendo o primeiro considerando apenas a covariável yschool e o segundo yschool e abdpain. Indique qual das afimativas abaixo é verdadeira.

In [ ]:
# Ajustando o modelo 1
fit51 = coxph(Surv(tempos,cens) ~ yschool, data=dados, x = T)
summary(fit51)

Call:
coxph(formula = Surv(tempos, cens) ~ yschool, data = dados, x = T)

  n= 877, number of events= 347 

            coef exp(coef) se(coef)      z Pr(>|z|)    
yschool -0.16058   0.85165  0.03294 -4.874 1.09e-06 ***
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

        exp(coef) exp(-coef) lower .95 upper .95
yschool    0.8517      1.174    0.7984    0.9085

Concordance= 0.567  (se = 0.016 )
Likelihood ratio test= 23.95  on 1 df,   p=1e-06
Wald test            = 23.76  on 1 df,   p=1e-06
Score (logrank) test = 23.5  on 1 df,   p=1e-06


In [ ]:
# Ajustando o modelo 2
fit52 = coxph(Surv(tempos,cens) ~ yschool + abdpain, data=dados, x = T)
summary(fit52)

Call:
coxph(formula = Surv(tempos, cens) ~ yschool + abdpain, data = dados, 
    x = T)

  n= 877, number of events= 347 

            coef exp(coef) se(coef)      z Pr(>|z|)    
yschool -0.16229   0.85020  0.03314 -4.897 9.75e-07 ***
abdpain  0.30985   1.36323  0.14634  2.117   0.0342 *  
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

        exp(coef) exp(-coef) lower .95 upper .95
yschool    0.8502     1.1762    0.7967    0.9073
abdpain    1.3632     0.7336    1.0233    1.8161

Concordance= 0.577  (se = 0.016 )
Likelihood ratio test= 28.13  on 2 df,   p=8e-07
Wald test            = 27.91  on 2 df,   p=9e-07
Score (logrank) test = 27.73  on 2 df,   p=1e-06


a) A inclusão da covariável npartner no segundo modelo torna a covariável yschool significativa.

In [ ]:
# Ajustando o modelo 2, incluindo a variável npartner
fit522 = coxph(Surv(tempos,cens) ~ yschool + abdpain + npartner, data=dados, x = T)
summary(fit522)

Call:
coxph(formula = Surv(tempos, cens) ~ yschool + abdpain + npartner, 
    data = dados, x = T)

  n= 877, number of events= 347 

             coef exp(coef) se(coef)      z Pr(>|z|)    
yschool  -0.16328   0.84935  0.03324 -4.912 9.02e-07 ***
abdpain   0.30038   1.35038  0.14675  2.047   0.0407 *  
npartner  0.04937   1.05061  0.04990  0.989   0.3225    
---
Signif. codes:  0 ‘***’ 0.001 ‘**’ 0.01 ‘*’ 0.05 ‘.’ 0.1 ‘ ’ 1

         exp(coef) exp(-coef) lower .95 upper .95
yschool     0.8494     1.1774    0.7958    0.9065
abdpain     1.3504     0.7405    1.0128    1.8004
npartner    1.0506     0.9518    0.9527    1.1586

Concordance= 0.575  (se = 0.017 )
Likelihood ratio test= 29.02  on 3 df,   p=2e-06
Wald test            = 28.73  on 3 df,   p=3e-06
Score (logrank) test = 28.58  on 3 df,   p=3e-06


**Falsa:** a covariável yschool já era significativa antes da inclusão de npartner.


b) A inclusão da covariável npartner no segundo modelo indica que um aumento no número de parceiros está associado a uma redução no risco de reinfecção para a cováriavel yschool, já que o coeficiente era -0.16047 e passou a ser -0.16161.

**Falsa:** tal diferença está mais associada ao ajuste do modelo do que no impacto da inclusão da covariável npartner sobre yschool.


c) O coeficiente de yschool no primeiro modelo indica que um aumento no tempo de escolaridade está associado a um aumento no risco de reinfecção.

**Falsa:** o coeficiente negativo indica que aumento no tempo de escolaridade está associado à diminuição do risco de reinfecção, o hazard ratio de 0.85165 indica que cada unidade adicional no tempo de escolaridade tem um impacto sob a diminuição do risco de reinfecção de aproximadamente 15%.

d) No segundo modelo, o coeficiente de yschool continua significativo, enquanto o coeficiente de npartner não é estatisticamente significativo.

**Verdadeira:** considerando um nível de significância de 0.05, temos que o valor-p = 9.02e-07 de yschool nos traz evidências em favor de sua significância estatística, enquanto o de npartner(0.3225) nos traz evidências contra a covariável.